rpmlb/builder/koji.py

```diff
@@ -0,0 +1,173 @@
+"""Builder interface for the Koji build service."""
+
+import re
+from pathlib import Path
+from typing import Iterator, Optional, Sequence
+
+from ..utils import run_cmd, run_cmd_with_capture
+from ..work import Work
+from .base import BaseBuilder
+
+
+class KojiBuilder(BaseBuilder):
+    """Builder interface for the Koji build service."""
+
+    DEFAULT_TARGET_FORMAT = 'sclo{epel}-{collection}-rh-el{epel}'
+
+    def __init__(self):
+        """Initialize the slots"""
+
+        self.collection = None  #: The collection to build
+        self.epel = None  #: Version of EPEL to build for
+        self.profile = None  #: The koji configuration profile to use
+        #: The format string for construction of the actual template name
+        self.target_format = type(self).DEFAULT_TARGET_FORMAT
+
+    @property
+    def base_command(self) -> Sequence[str]:
```

Comment:
There is no such thing as protected method in Python. There is only private with two underscores, but people are not using those (except for magic `__mehtods__` like this). One underscore in front of a method does not mean it's private nor protected, it does not mean anything. It's merely a convention. And explicit meaning of such underscore may vary form project to projet. As I read it (by experience) it usually means: _This is internal, it might change, it my go away, don't use it, don't override it in subclasses._

The stuff you talk about is mostly about languages where private and protected have meanings.

The Java example makes sense, you cannot access `list_a.some_method_a()` because you declared it a `List`. It has nothing to do with Python, because in python, you don't declare things as a specific type. Nevertheless nothing forbids you form creating such method and using it if the object was not declared as a `List`.

Look: a Vector in Java has a public method [capacity()](https://docs.oracle.com/javase/7/docs/api/java/util/Vector.html#capacity()) that is not specified in List interface or AbscractList base class. It's just there because *it is useful*.

Let me look at the Python example instead: Noone is calling a method only specified on KojiBuilder in a scope where it might be a different type of builder. The last line of the Python example you've provided is indeed wrong, no argue in that. But that doesn't mean in any way that simply presence of such method in the KojiBuilder class is wrong.

Polymorphism, Loose coupling, Encapsulation, nothing in there forbids you form extending the class interface in a subclass.

So let me repeat my question: **What is a benefit of not extending the API of a subclass?** Please be specific to our use case in KojiBuilder. Or, if you want to talk generally about OOP patterns, please show me where any of those terms you've used forbids to extend an API of a subclass.